File: app/src/main/java/com/aaa/breakout/GameSurfaceView.java
package com.aaa.breakout;

import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.PorterDuff;
import android.graphics.Rect;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.SurfaceHolder;
import android.view.SurfaceView;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import java.util.ArrayList;
import java.util.List;

public class GameSurfaceView extends SurfaceView implements Runnable, SurfaceHolder.Callback {
    private boolean isRunning = false;
    private Thread gameThread = null;
    private SurfaceHolder mHolder = null;
    private float eventX = 0f;
    private boolean mSoundFlg = false;

    /* Viewを継承したときのお約束 */
    public GameSurfaceView(@NonNull Context context, @Nullable AttributeSet attrs) { this(context, attrs, 0); }
    public GameSurfaceView(@NonNull Context context, @Nullable AttributeSet attrs, int defStyleAttr) { super(context, attrs, defStyleAttr); }

    @Override
    public void surfaceCreated(@NonNull SurfaceHolder holder) {
        /* ↓kotlinの時はこれがないと描画が見えなかった。 */
//      getHolder().setFormat(PixelFormat.TRANSLUCENT);
    }

    @Override
    public void surfaceChanged(@NonNull SurfaceHolder holder, int format, int width, int height) {
        /* 描画開始 */
        PhaseManager.init(getContext(), mSoundFlg, width, height);
        mHolder = holder;
        isRunning = true;
        gameThread = new Thread(this);
        gameThread.start();
        /* ジェスチャバックを無効化 */
        List<Rect> rectlist = new ArrayList<>();
        rectlist.add(new Rect(0, height-650, width, height-450));
        rectlist.add(new Rect(0, height-450, width, height-250));
        rectlist.add(new Rect(0, height-250, width, height-50));
        setSystemGestureExclusionRects(rectlist);
    }

    @Override
    public void surfaceDestroyed(@NonNull SurfaceHolder holder) {
        isRunning = false;
        try {
            gameThread.join();
        } catch (InterruptedException e) {
            Log.d("aaaaa", "error2!!", e);
        }
        gameThread = null;
    }

    @Override
    public void run() {
        while(isRunning) {
            try {
                final int FRAMERATE = 15;
                Thread.sleep(FRAMERATE);
            } catch (InterruptedException e) {
                Log.e("aaaaa", "error!!", e);
            }

            Canvas canvas = mHolder.lockCanvas();
            if(canvas == null) continue;
            /* 一旦全消去 */
            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR);

            PhaseBase drawPhase = PhaseManager.getPhase();
            drawPhase.update(eventX);
            drawPhase.draw(canvas);

            mHolder.unlockCanvasAndPost(canvas);
        }
    }

    @Override
    public boolean performClick() {
        return super.performClick();
    }

    @Override
    public boolean onTouchEvent(MotionEvent event) {
        super.onTouchEvent(event);
        if(event.getAction() == MotionEvent.ACTION_DOWN || event.getAction() == MotionEvent.ACTION_MOVE) {
            eventX = event.getX();
            performClick();
            return true;
        }
        return false;
    }

    public void setSoundFlg(boolean soundFlg) {
        mSoundFlg = soundFlg;
    }
}

File: app/src/main/java/com/aaa/breakout/GameSurfaceView.kt
package com.aaa.breakout

import android.R.attr
import android.content.Context
import android.graphics.Color
import android.graphics.PorterDuff
import android.graphics.Rect
import android.util.AttributeSet
import android.util.Log
import android.view.MotionEvent
import android.view.SurfaceHolder
import android.view.SurfaceView
import com.aaa.breakout.PhaseManager.init
import com.aaa.breakout.PhaseManager.phase

const val FRAMERATE = 15

class GameSurfaceView: SurfaceView, Runnable, SurfaceHolder.Callback {
    private var isRunning = false
    private lateinit var gameThread: Thread
    private lateinit var mHolder: SurfaceHolder
    private var eventX = 0f
    private var mSoundFlg = false

    /* Viewを継承したときのお約束 */
    constructor(context: Context): super(context)
    constructor(context: Context, attrs: AttributeSet): super(context, attrs)
    constructor(context: Context, attrs: AttributeSet, defStyleAttr: Int): super(context, attrs, defStyleAttr)

    override fun surfaceCreated(holer: SurfaceHolder) {
        /* ↓kotlinの時はこれがないと描画が見えなかった。 */
//        getHolder().setFormat(PixelFormat.TRANSLUCENT);
    }

    override fun surfaceChanged(holder: SurfaceHolder, format: Int, width: Int, height: Int) {
        /* 描画開始 */
        init(context, mSoundFlg, width, height)
        mHolder = holder
        isRunning = true
        gameThread = Thread(this)
        gameThread.start()

        /* ジェスチャバックを無効化 */
        systemGestureExclusionRects = ArrayList<Rect>().apply {
            add(Rect(0, attr.height - 650, attr.width, attr.height - 450))
            add(Rect(0, attr.height - 450, attr.width, attr.height - 250))
            add(Rect(0, attr.height - 250, attr.width, attr.height - 50))
        }
    }

    override fun surfaceDestroyed(holder: SurfaceHolder) {
        isRunning = false
        try {
            gameThread.join()
        } catch (e: InterruptedException) {
            Log.d("aaaaa", "error2!!", e)
        }
    }

    override fun run() {
        while (isRunning) {
            try {
                Thread.sleep(FRAMERATE.toLong())
            } catch (e: InterruptedException) {
                Log.e("aaaaa", "error!!", e)
            }

            val canvas = mHolder.lockCanvas() ?: continue
            /* 一旦全消去 */
            canvas.drawColor(Color.TRANSPARENT, PorterDuff.Mode.CLEAR)

            val drawPhase = phase
            drawPhase.update(eventX)
            drawPhase.draw(canvas)

            mHolder.unlockCanvasAndPost(canvas)
        }
    }

    override fun performClick(): Boolean {
        super.performClick()
        return true
    }

    override fun onTouchEvent(event: MotionEvent): Boolean {
        super.onTouchEvent(event)
        if (event.action == MotionEvent.ACTION_DOWN || event.action == MotionEvent.ACTION_MOVE) {
            eventX = event.x
            performClick()
            return true
        }
        return false
    }

    fun setSoundFlg(soundFlg: Boolean) {
        mSoundFlg = soundFlg
    }
}